### ISSUE TITLE
WorkloadEntry to support healthStatus and ServiceEntry will exclude Unhealthy WLEs

### ISSUE THREAD

**Issue by shonecyx** [istio/istio#37073]
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)
This is related with https://github.com/istio/istio/issues/36865

Introduce a new spec filed `healthStatus` in WLE and Healthy/Unknown ones will be selected by SE. For backward compatible, default will be Unknown and only `Unhealthy` WLEs will not be selected.
```
apiVersion: networking.istio.io/v1alpha3
kind: WorkloadEntry
metadata:
  name: details-svc
spec:
  serviceAccount: details-legacy
  address: 2.2.2.2
  healthStatus: Healthy/Unknown/Unhealthy
  labels:
    app: details-legacy
```

According the upstream doc, WorkloadEntry enables operators to describe the properties of a single non-Kubernetes workload such as a VM or a bare metal server as it is onboarded into the mesh, which means WLE actually turns the non-Kubernetes VMs into a Pod.

In k8s, we have below resources correlated:
```
Service ---> Endpoint ----> Pod
```
Since there is readinessProbe in Endpoint,  Istio discovery service will eventually discover the healthy pods since they will be in not ready addresses in Endpoint, which means the pod readiness status will eventually be reflected to the Envoy LbEndpoint in the upstream cluster.

And Istio introduced few other CRDs to represent the non-k8s workloads:
```
ServiceEntry ---> WorkloadGroup ---> WorkloadEntry
```
However currently if the WorkloadEntry is down, there is no such mechanism to unselect that WLE from the SE since there is no such field to store the healthStatus info and suppose this is a very common request for the WLE traffic management.

Although there is `ReadinessProbe` in WorkloadGroup, it doesn't work as k8s which is the kubelet to do the readinessProbe against Pod and populate the status to Endpoint and the readinessProbe is coupled with WorkloadGroup so we won't be able to on demand mark down a WLE. 

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by howardjohn** [comment]
> Although there is ReadinessProbe in WorkloadGroup, it doesn't work as k8s which is the kubelet to do the readinessProbe against Pod and populate the status to Endpoint and the readinessProbe is coupled with WorkloadGroup so we won't be able to on demand mark down a WLE.

I don't understand how this is different from Pods. In Pod, kubelet does readiness probe and marks it as unhealthy. A user cannot explicitly trigger 'unhealthy' other than making it fail the probe.

In WE, istio-agent does readiness probe and marks it as unhealthy. A user cannot explicitly trigger 'unhealthy' other than making it fail the probe.

So what is the difference? Why is this needed for WE but not Pod?

**Comment by shonecyx** [comment]
@howardjohn  One significant difference is for the k8s service without selector, we can have external readiness probe/controller to maintain the healthy and unhealthy addresses, so the service/endpoint is decoupled with the readiness probe and fully leverage on the ip address health status in the endpoint and only the ready addresses will take traffic. With this design, we'll have the flexibility to manage the addresses and endpoint, we can either fail the readiness probe or just change the address to not ready without actually deleting it. However we don't have that capability/flexibility with WLE and SE.
In another word, the requirement here is to enable SE to only select healthy WLE and doesn't matter if there is readiness probe or not, just like the k8s Service without selector to only select ready addresses.

**Comment by shonecyx** [comment]
@howardjohn  Sorry for late reply.

The consideration for using WLE instead of Endpoint is because the semantic of WLE is to represent non-k8s workloads, it could be VM or VIP which is exactly our case.

From spec perspective, we would like to capture the intention to not send traffic to a specific endpoint, it could be vm/vip and there are two options like I mentioned in another issue:
1. Support to set the WLE weight to be really 0 which means istiod will either mark that endpoint as unhealthy or just delete it from envoy cluster. 
2. Introduce a new filed healthStatus in WLE spec which will mark the LbEndpoint to be not healthy, just like the not ready addresses in k8s Endpoint.

Both options will work for us. Deleting the WLE is a workaround to achieve this but the intention of `delete` and `disable` is different.

**Comment by howardjohn** [comment]
What I still don't understand is:
1, Why does existing health mechanism on WE not satisfying use case
2. Given pod does not enable this, and WE is equivalent to Pod, why is this needed? You said you already manage this with a Service without selector for Pod - why not just do that for WE?

**Comment by shonecyx** [comment]
@howardjohn 
>1, Why does existing health mechanism on WE not satisfying use case

Exiting health mechanism is the active health through istio-agent for pod as I went through the BP [Virtual Machine Health Checking](https://docs.google.com/document/d/1G4s_PrmvkwCKoH_kXVeSUAszsaj3ZfpIlraSdeTSPKU/edit#heading=h.ejtjwkmtydnh). However our case the WLE represents a VIP, it could be a hardware LB VIP or software LB(IPVS) VIP, for that case there is no istio-agent to do the health check.  Another requirement is that, even if we can have another external controller to monitor the WLE health status,  there is a scenario that we want to do VIP exit even if it's healthy for the region exit case.
>2. Given pod does not enable this, and WE is equivalent to Pod, why is this needed? You said you already manage this with a Service without selector for Pod - why not just do that for WE?

As I mentioned, it's VIP in our case and we'd like to decouple the HC with istio-agent. And actually there are two use cases:
1. Active health check against the VIP,  we can refer to Contour's design https://projectcontour.io/docs/v1.10.0/config/health-checks/ which works better for VIP case.
2. On demand disable an Endpoint(WLE/VIP) to exit traffic no matter it's healthy or not. If we can't introduce the `healthStatus` in WLE spec, we are going to put the intention into an annotation and istiod need to mark the corresponding LbEndpoint as unhealthy based on this annotation. If we make this change, can we contribute it back to upstream?
```
      - apiVersion: networking.istio.io/v1beta1
        kind: WorkloadEntry
        metadata:
           annotations:
             traffic.network.istio.io/disablePorts: 80,443 or * for all ports
          creationTimestamp: null
          name: https-foo-we
        spec:
          address: foo-az01.example.com
          labels:
            app: foo
          weight: 50
          Locality: region01/az01
        status: {}
```

**Comment by howardjohn** [comment]
For case 1 I don't mean using Istio-agent to do health checks I mean to set the Healthy status condition your self. We already have one way to represent health so I do not want to have two.

**Comment by shonecyx** [comment]
> For case 1 I don't mean using Istio-agent to do health checks I mean to set the Healthy status condition your self. We already have one way to represent health so I do not want to have two.

I see, then it works for case #1 and still need to put the annotation `WorkloadEntryHealthCheckAnnotation` to WLE and manually update the WLE status but actually no readinessProbe.
```
// isHealthy checks that the provided WorkloadEntry is healthy. If health checks are not enabled,
// it is assumed to always be healthy
func isHealthy(cfg config.Config) bool {
	if parseHealthAnnotation(cfg.Annotations[status.WorkloadEntryHealthCheckAnnotation]) {
		// We default to false if the condition is not set. This ensures newly created WorkloadEntries
		// are treated as unhealthy until we prove they are healthy by probe success.
		return status.GetBoolConditionFromSpec(cfg, status.ConditionHealthy, false)
	}
	// If health check is not enabled, assume its healthy
	return true
}
```
And there is still another issue to address, how to disable WLE on specific port and capture the disabled result into WLE status? The port information is in SE and currently there is no port information in WLE status.

**Comment by howardjohn** [comment]
If you want to disable on a specific port I would encourage you to make multiple distinct WE for each port. Having a per-port status is likely not something that generalizes beyond your use case so it would be hard to fit into the api

**Comment by shonecyx** [comment]
> If you want to disable on a specific port I would encourage you to make multiple distinct WE for each port. Having a per-port status is likely not something that generalizes beyond your use case so it would be hard to fit into the api

I just tried out to disable the WLE:
```
apiVersion: networking.istio.io/v1beta1
kind: WorkloadEntry
metadata:
  annotations:
    proxy.istio.io/health-checks-enabled: "true"
  name: foo
  namespace: foo-ns
spec:
  address: 10.0.0.1
  labels:
    app: foo
  locality: region01/az01
  weight: 50
status:
  conditions:
  - lastProbeTime: "2019-12-26T22:06:34Z"
    lastTransitionTime: "2019-12-26T22:06:34Z"
    message: 'Code 500: VIP Unhealthy'
    reason: unhealthy
    status: "False"
    type: Healthy

```
Before add the annotation/status:
```
istioctl pc endpoint -n isito-staging istio-ingressgateway-6cbc9779fc-422kw.istio-staging | grep foo
10.0.0.1:8082               HEALTHY     OK                outbound|18080||foo.example.com
10.0.0.1:8082               HEALTHY     OK                outbound|8082||foo.example.com
10.0.0.2:8082               HEALTHY     OK                outbound|18080||foo.example.com
10.0.0.2:8082               HEALTHY     OK                outbound|8082||foo.example.com
```
After add the annotation/status:
```
istioctl pc endpoint -n isito-staging istio-ingressgateway-6cbc9779fc-422kw.istio-staging | grep foo
10.0.0.2:8082               HEALTHY     OK                outbound|18080||foo.example.com
10.0.0.2:8082               HEALTHY     OK                outbound|8082||foo.example.com
```
We can see the 10.0.0.1 WLE got removed from the endpoint with both port 8082/18080. BTW here is the SE spec:
```
- apiVersion: networking.istio.io/v1beta1
  kind: ServiceEntry
  metadata:
    creationTimestamp: null
    name: foo
  spec:
    hosts:
    - foo.example.com
    ports:
    - name: http-8080
      number: 18080
      protocol: HTTP
    - name: https-8082
      number: 8082
      protocol: HTTPS
    resolution: DNS
    workloadSelector:
      labels:
        app: foo
  status: {}
```
This is very close to what we need and the only concern is that we are not able to disable WLE at port level since by default it applies for all ports.

Create multiple WLE for each port is a workaround but we'll also need to create multiple SE to select the WLE and it will significantly increase the SE/WLE object number, considering we could have thousands of WLE it might also case performance issue.

Shall we introduce the annotation `traffic.network.istio.io/disablePorts: 80,443 or * for all ports` in WLE to only disable specific port?  For example if we only disable `8082` for `10.0.0.1`,  the expected output is:
```
istioctl pc endpoint -n isito-staging istio-ingressgateway-6cbc9779fc-422kw.istio-staging | grep foo
10.0.0.1:8082               HEALTHY     OK                outbound|18080||foo.example.com
10.0.0.2:8082               HEALTHY     OK                outbound|18080||foo.example.com
10.0.0.2:8082               HEALTHY     OK                outbound|8082||foo.example.com
```

**Comment by shonecyx** [comment]
I went through the code and the change is servicediscovery to honor the WLE annotation `traffic.network.istio.io/disablePorts: 80,443 or * for all ports` and delete the corresponding ServiceInstance with the port:
https://github.com/istio/istio/blob/release-1.12/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go#L229. 

Through this way we can avoid creating duplicated SE/WLE per port and also support to disable WEL at port level.

**Comment by howardjohn** [comment]
Ah I thought you could do this:
```yaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  creationTimestamp: null
  name: foo
spec:
  hosts:
  - foo.example.com
  ports:
  - name: http-8080
    number: 18080
    protocol: HTTP
  - name: https-8082
    number: 8082
    protocol: HTTPS
  resolution: DNS
  workloadSelector:
    labels:
      app: foo
---
apiVersion: networking.istio.io/v1beta1
kind: WorkloadEntry
metadata:
  name: foo-1
spec:
  ports:
    http-8080: 18081
  address: 10.0.0.1
  labels:
    app: foo
  locality: region01/az01
  weight: 50

```

To make WE only apply to one port. But it does not.

I am not in favor of anything like `disablePorts`. Health is at the workload level, not the port level in Kubernetes and we are trying to mirror that.

**Comment by shonecyx** [comment]
Thanks @howardjohn . We'll take the approach to create SE/WLE per port for our case. BTW two more questions:
1. I tried out the ports in WLE `http-8080: 18081` and after I added the port map, data plane is not working. Looks like there is a bug?
2. The annotation `proxy.istio.io/health-checks-enabled: "true"` was introduced for VM health check through pilot agent, I wonder how can we support active health check for VIP case?  Will your previous commit help for that case:  https://github.com/howardjohn/istio/commit/4519a02e05cdd242cca7f4b53f7053b7489e88d2. Or just leverage EnvoyFilter for that case？
